File: src/types/practice.ts
export type AttemptStatus = 'recorded' | 'analyzed';
export type AnalysisStatus = 'not_analyzed' | 'completed' | 'failed';
export type MistakeType = 'missing' | 'extra' | 'substituted';

/**
 * A single recorded/transcribed word discrepancy between what the ayah
 * actually says and what Whisper transcribed from the recording.
 * `ayahNumber` is null only for 'extra' (a recited word with no
 * reference counterpart to attribute it to).
 */
export interface PracticeMistake {
  mistakeType: MistakeType;
  ayahNumber: number | null;
  referenceWord: string | null;
  transcribedWord: string | null;
}

/**
 * A single practice recording. Phase 14 added real, on-demand recitation
 * analysis (`analysisStatus` starts 'not_analyzed' until the student taps
 * "Analyze recitation") — word-level transcription diffing, not
 * pronunciation/fluency/Tajweed scoring. This app still never reports a
 * `pronunciationScore`/`fluencyScore`-style number it can't actually
 * compute; there's deliberately no such field on this type at all.
 */
export interface PracticeAttempt {
  id: string;
  surahNumber: number;
  surahName: string;
  fromAyah: number;
  toAyah: number;
  recordedAt: string; // ISO datetime
  durationSeconds: number;
  status: AttemptStatus;
  pace: PaceEstimate;
  /** Set only if the recording was actually uploaded and saved server-side (Phase 10). */
  audioUrl?: string | null;

  analysisStatus: AnalysisStatus;
  analysisError?: string | null;
  transcribedText?: string | null;
  matchedWordCount?: number | null;
  totalWordCount?: number | null;
  mistakes: PracticeMistake[];

  /**
   * Phase 26: true only for a locally-synthesized placeholder returned
   * while offline — this attempt was queued, not yet accepted by the
   * server, and its `id` isn't a real server id. Never set on a real
   * API response.
   */
  isPendingSync?: boolean;
}

/**
 * A local, non-AI pace signal: how the recording's duration compares to a
 * rough expected range for the ayah count, based on typical tarteel pace.
 * This is NOT recitation analysis — it can't see pronunciation, fluency, or
 * mistakes. It exists only so Practice Mode has *something* useful before
 * Phase 10's real scoring lands, without pretending to be more than it is.
 */
export interface PaceEstimate {
  expectedSecondsRange: [number, number];
  actualSeconds: number;
  withinExpectedRange: boolean;
}

File: backend/app/schemas/practice.py
from datetime import datetime
from typing import Literal

from app.schemas.base import CamelModel


class PaceOut(CamelModel):
    expected_seconds_range: tuple[float, float]
    actual_seconds: float
    within_expected_range: bool


MistakeType = Literal["missing", "extra", "substituted"]
AnalysisStatus = Literal["not_analyzed", "completed", "failed"]


class PracticeMistakeOut(CamelModel):
    mistake_type: MistakeType
    ayah_number: int | None
    reference_word: str | None
    transcribed_word: str | None


class PracticeAttemptOut(CamelModel):
    id: str
    surah_number: int
    surah_name: str
    from_ayah: int
    to_ayah: int
    recorded_at: datetime
    duration_seconds: float
    status: str
    audio_url: str | None = None
    pace: PaceOut

    # Phase 14 — recitation analysis. `analysis_status` is 'not_analyzed'
    # until explicitly requested (POST .../analyze); pronunciation/fluency
    # scores are deliberately absent from this model entirely — this app
    # never reports a number it can't actually compute (see
    # services/arabic_text.py for exactly what word-diffing can and can't
    # tell you).
    analysis_status: AnalysisStatus = "not_analyzed"
    analysis_error: str | None = None
    transcribed_text: str | None = None
    matched_word_count: int | None = None
    total_word_count: int | None = None
    mistakes: list[PracticeMistakeOut] = []


class PracticeAttemptCreate(CamelModel):
    surah_number: int
    surah_name: str
    from_ayah: int
    to_ayah: int
    duration_seconds: float
    expected_min_seconds: float
    expected_max_seconds: float
    # Phase 26: a client-generated UUID so a queued offline write, synced
    # later (possibly retried if the sync itself gets interrupted), never
    # creates a duplicate attempt — see services/idempotency.py.
    client_mutation_id: str | None = None
